app/services/contacts_service.rb

```diff
@@ -1,61 +1,63 @@
-# Provides service for calculating Contact objects out of AgentStatus data that corresponds to answered contacts
+# Calculates various statistics data based on Contact objects.
 class ContactsService
-  def contacts_for_team(team_name, starttime, endtime)
-    convert_to_contacts(contact_statuses(team_name, starttime, endtime))
+  def initialize(team_name, start_time, end_time)
+    @contacts = Contact.joins(service: :team).where(teams: { name: team_name }, arrived_in_queue: start_time..end_time)
   end
 
-  def answered_calls(team_name, start_time, end_time)
-    contact_statuses(team_name, start_time, end_time).count
+  def num_answered_calls
+    answered_contacts.count
   end
 
-  def average_call_duration(team_name, start_time, end_time)
-    average_duration(contact_statuses(team_name, start_time, end_time))
+  def average_call_duration
+    average_duration(answered_contacts, 'answered', 'call_ended')
   end
 
-  def average_after_call_duration(team_name, start_time, end_time)
-    average_duration(statuses(team_name, start_time, end_time, 'Jälkikirjaus'))
+  def average_after_call_duration
+    contacts = @contacts.where(contact_type: 'PBX').where.not(call_ended: nil, answered: nil, service_id: 120, handling_ended: nil)
+    average_duration(contacts, 'call_ended', 'handling_ended')
   end
 
-  def calls_by_hour(team_name, start_time, end_time)
-    gmt_offset = start_time.gmt_offset
-    select = "EXTRACT(HOUR FROM agent_statuses.created_at + '#{gmt_offset} seconds') AS hour, COUNT(*) AS count"
-    data = contact_statuses(team_name, start_time, end_time).select(select).group('hour')
+  def answered_percentage
+    answered = num_answered_calls
+    missed = num_missed_calls
+    return 100.0 if missed == 0
+    (answered.to_f / (answered + missed) * 100).round(1)
```

Comment:
Division by zero can happen because of `answered + missed`. Result is infinity but JSON doesn't support it. Use `nil` instead?